presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java

```diff
@@ -96,7 +96,6 @@ public JoinNode(@JsonProperty("id") PlanNodeId id,
                 .addAll(right.getOutputSymbols())
                 .build();
         checkArgument(new HashSet<>(inputSymbols).containsAll(outputSymbols), "Left and right join inputs do not contain all output symbols");
-        checkArgument(!isCrossJoin() || inputSymbols.equals(outputSymbols), "Cross join does not support output symbols pruning or reordering");
```

Comment:
+1 I know you moved the check here: https://github.com/prestodb/presto/pull/9585/commits/4bb1faf25a977c609634f607362d38aa8b216216#diff-e5a018f043de654e40baeb9ce677d4aaR1573,
but it's better if it stays in `JoinNode` since it will fail earlier and exactly in place where join is incorrectly constructred.

Reply:
@findepi , @sopel39 , when I changed JoinNode#isCrossJoin to remove `&& !filter.isPresent()` this check started failing. It failed while running a query from `TestLogicalPlanner#testLeftConvertedToInnerInequalityJoinNoEquiJoinConjuncts`.

```
SELECT 1 
FROM orders o LEFT JOIN lineitem l ON o.orderkey < l.orderkey 
WHERE l.orderkey IS NOT NULL
```

While planning this query, Presto first creates JoinNode with `type = LEFT and filter = 'o.orderkey < l.orderkey'`. Then `PredicatePushDown.Rewriter#tryNormalizeToOuterToInnerJoin` changes join type to INNER and keeps the filter. At this point inputSymbols [orderkey, orderkey_0] don't match output symbols [orderkey_0]. Previous version of isCrossJoin() returned false (because of the filter), but new version returns true. Hence, the check fails. 

Finally, PredicatePushDown updates outputSymbols for the JoinNode to match inputSymbols and the check starts passing again:

```
    output = new JoinNode(
                        idAllocator.getNextId(),
                        node.getType(),
                        leftSource,
                        rightSource,
                        equiJoinClauses,
                        ImmutableList.<Symbol>builder()
                                .addAll(leftSource.getOutputSymbols())
                                .addAll(rightSource.getOutputSymbols())
                                .build(),
                        newJoinFilter,
                        node.getLeftHashSymbol(),
                        node.getRightHashSymbol(),
                        node.getDistributionType());
```

To keep the check in the JoinNode constructor, we'd need to ensure that all intermediate versions of JoinNode pass that check. It looks like this check was passing by accident before (why would isCrossJoin return false for a cross-join with a filter?). What do you think?

Reply:
@mbasmanova very good analysis. I am concerned about `PredicatePushDown` producing invalid `JoinNode` and then, at some other point, making it valid again. Functionally this is sufficient, but I would rather not be a person that debugs this code at some later time.

Reply:
> why would isCrossJoin return false for a cross-join with a filter?
This wasn't precise very much, but `isCrossJoin` actually meant join that is executed by nested loop operator, so this was implementation leaking to planner. Technically, it would be best if there is generic `JoinNode` and implementation nodes, e.g: `NestedLoopJoinNode`.

However, as bad as it was, we probably shouldn't rely that particular rule execution order will fix the plan for us. It it is possible, we should try to have valid nodes during entire planning process.